项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入NumPy和Pandas数据库，用Pandas的`read_csv`函数读取原始数据集'e_commerce.csv'，使其转换成DataFrame格式，并赋值给变量`df`。

展示数据集的前5行和后5行。

In [1]:
import numpy as np
import pandas as pd

In [2]:
df= pd.read_csv('e_commerce.csv')
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

## 评估数据

在这一部分，我将对上一部分所建立的`df`这个DataFrame所包含的数据进行评估

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 一、评估数据整洁度

利用DataFrame的info方法和sample方法，分别提供数据的概况信息和实际数据，来评估数据的整洁度

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


列名符合每列只能表示一个变量的整洁性规则

In [4]:
df.sample(5)

InvoiceNo StockCode                     Description  Quantity  \
374695    569401     22111    SCOTTIE DOG HOT WATER BOTTLE         3   
354899    567901     21231  SWEETHEART CERAMIC TRINKET BOX         1   
301989    563367     21668  RED STRIPE CERAMIC DRAWER KNOB        12   
86547     543555    85099F            JUMBO BAG STRAWBERRY        10   
535569    581217     22212      FOUR HOOK  WHITE LOVEBIRDS         3   

            InvoiceDate  UnitPrice  CustomerID         Country  
374695   10/4/2011 8:56       4.95     15111.0  United Kingdom  
354899  9/22/2011 16:28       2.46         NaN  United Kingdom  
301989   8/16/2011 8:58       1.25     16133.0  United Kingdom  
86547   2/10/2011 10:18       1.95     17686.0  United Kingdom  
535569   12/8/2011 9:20       4.13         NaN  United Kingdom

从抽样的10行数据来看，符合'每列是一个变量，每行是一个观察值，每个单元格是一个值'这三个标准。具体来看每行是关于某商品的一次交易数据，每列是关于交易的各个变量，因此不存在结构性问题。

### 二、评估数据干净度

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


从输出结果来看，数据共有541909条观察值。

`Description`和`CustomerID`列存在缺失值。

`CustomerID`列数据类型应为object，`InvoiceDate`数据类型应为datetime，应当进行数据类型转换

#### (一)、评估丢失数据

初步统计各列缺失值的数量

In [6]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

在了解`Description`存在缺失值后，根据条件提取出缺失观察值。

In [7]:
df[df.Description.isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

有1454条交易数据缺失`Description`变量值

从输出结果来看，这些缺失`Description`的交易数据，`UnitPrice`都为0，且`CustomerID`同时缺失。为了验证猜想，我们增加筛选条件，看是否存在`Description`变量缺失，且`UnitPrice`不为0的数据

In [8]:
df[(df.Description.isnull()) & (df.UnitPrice != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

验证猜想，输出结果为0条，说明缺失`Description`变量值的数据，同时也不具备有效的`UnitPrice`值。

`Description`表示产品名称，`UnitPrice`表示产品单价，都是进行后续数据分析的关键变量。如果他们同时缺失，我们认为这些数据无法提供有效含义，因此这些后续可以被删除。

变量`CustomerID`同样存在缺失值，因此也根据条件提取出缺失观察值。

In [9]:
df[df.CustomerID.isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

有135080条交易数据缺失`CustomerID`变量值

从输出结果来看，缺失`CustomerID`变量值不一定缺失其他变量值，并且`CustomerID`不是后续分析畅销商品的关键变量，因此保留此变量为空的观察值。

#### (二)、评估重复数据

根据数据变量含义来看，虽然`InvoiceNo`、`StockCode`和`CustomerID`都是唯一标识符，但是一次交易可以包含多件商品，因此`InvoiceNo`可以存在重复；不同交易可以选择同件商品，因此`InvoiceNo`可以存在重复；每个客户可以产生不同交易，因此`CustomerID`可以存在重复。但是数据所有变量同时重复时，应被看作重复数据。而不是某个顾客一次买了多件同一商品，毕竟这样应该是quantity变量的值为多个。

In [11]:
df[df.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

从输出结果来看，数据中存在5268条重复数据，应在数据清洗环节删除重复数据。

#### (三)、 评估不一致数据 

从数据变量含义来看，只能判断`Country`变量是否存在不一致数据

In [12]:
df.Country.value_counts()

United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68
European

从`Country`变量值来看，此处U.K., UK和United Kingdom都表示英国，United States和USA都表示美国。这些数据是不一致数据，在后续清理步骤需要把数值都替换成统一的

#### (四)、评估无效/错误数据

In [14]:
df.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

Quantity和UnitPrice列均存在无效数据，接下来筛选出Quantity和UnitPrice为负数的观察值，来进一步观察这些数据的特点

In [15]:
df[df['Quantity'] <= 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

可以发现Quantity为负数时，似乎InvoiceNo均为C开头，表示这笔交易被取消。

为了验证猜想，我们需要增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

In [16]:
df[(df['Quantity'] <= 0) & (df['InvoiceNo'].str[0] != 'C')]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

可以发现猜想并不准确，确实存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但进一步观察发现，这些筛选出的观察值UnitPrice均为0，因此我们增加筛选条件，进一步验证

In [17]:
df[(df['Quantity'] <= 0) & (df['InvoiceNo'].str[0] != 'C') & (df['UnitPrice'] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，证明当Quantity为负数的时候，观察值满足以下条件之一：

InvoiceNo以C开头，表示订单被取消；

UnitPrice为0.说明单价0英镑

这些交易数据不是有效交易数据，不在后续数据分析范围内，因此在数据清理阶段，将Quantity为负数的观察值删除

In [18]:
df.Quantity.sort_values()

540422   -80995
61624    -74215
225529    -9600
225530    -9600
4287      -9360
          ...  
421632     4800
74614      5568
502122    12540
61619     74215
540421    80995
Name: Quantity, Length: 541909, dtype: int64

In [19]:
df.UnitPrice.sort_values()

299984   -11062.06
299983   -11062.06
40984         0.00
52217         0.00
52262         0.00
            ...   
16356     13541.33
43703     16453.71
43702     16888.02
524602    17836.46
222681    38970.00
Name: UnitPrice, Length: 541909, dtype: float64

In [20]:
df.InvoiceNo.value_counts()

573585     1114
581219      749
581492      731
580729      721
558475      705
           ... 
554023        1
554022        1
554021        1
554020        1
C558901       1
Name: InvoiceNo, Length: 25900, dtype: int64

最后一条InvoiceNo，表示存在被取消的交易

In [21]:
df.StockCode.value_counts()

85123A    2313
22423     2203
85099B    2159
47566     1727
20725     1639
          ... 
21431        1
22275        1
17001        1
90187A       1
72759        1
Name: StockCode, Length: 4070, dtype: int64

In [22]:
df.InvoiceDate.value_counts()

10/31/2011 14:41    1114
12/8/2011 9:28       749
12/9/2011 10:03      731
12/5/2011 17:24      721
6/29/2011 15:58      705
                    ... 
9/28/2011 9:57         1
9/28/2011 9:45         1
3/16/2011 12:13        1
3/16/2011 12:22        1
3/16/2011 15:58        1
Name: InvoiceDate, Length: 23260, dtype: int64

## 清理数据

### 一、清理数据索引和列名

### 二、清理乱数据

### 三、清理脏数据

#### (一)、处理缺失数据

   因为此次数据分析的目的是挖掘畅销产品，因此Description是关键变量。此外，Description缺失值无法填充，且Description为缺失值的观察行关键变量UnitPrice的值也为0，因此删除Description的值缺失的行

In [23]:
df.dropna(subset=['Description'], inplace=True)
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     133626
Country             0
dtype: int64

#### (二)、删除重复数据

In [24]:
df.drop_duplicates(inplace=True)
df[df.duplicated()]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

#### (三)、处理不一致数据

In [25]:
df.replace(['U.K.', 'UK'], 'United Kingdom', inplace=True)
df.replace('USA', 'United States', inplace=True)
df.Country.value_counts()

United Kingdom          488846
Germany                   9480
France                    8541
EIRE                      8184
Spain                     2528
Netherlands               2371
Belgium                   2069
Switzerland               1994
Portugal                  1510
Australia                 1258
Norway                    1086
Italy                      803
Channel Islands            757
Finland                    695
Cyprus                     611
Sweden                     461
Unspecified                442
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     294
United States              291
China                      284
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58
Lebanon 

#### (四)、处理无效/错误数据

In [26]:
df.drop(df[df.Quantity <= 0].index, inplace=True)
df.Quantity.sort_values()

450813        1
445433        1
445432        1
445431        1
445430        1
          ...  
206121     4300
421632     4800
502122    12540
61619     74215
540421    80995
Name: Quantity, Length: 525462, dtype: int64

In [27]:
df.drop(df[df.UnitPrice <= 0].index, inplace=True)
df.UnitPrice.sort_values()

157195        0.001
361741        0.001
359871        0.001
279045        0.001
434038        0.040
            ...    
268028     4287.630
297723     4505.170
173382     8142.750
299982    11062.060
15017     13541.330
Name: UnitPrice, Length: 524878, dtype: float64

发现单价存在0.001英镑的，不合常理，应为错误数据

In [28]:
df.drop(df[df.UnitPrice == 0.001].index, inplace=True)
df.UnitPrice.sort_values()

301033        0.04
221726        0.04
363674        0.04
410140        0.04
407634        0.04
            ...   
268028     4287.63
297723     4505.17
173382     8142.75
299982    11062.06
15017     13541.33
Name: UnitPrice, Length: 524874, dtype: float64

In [29]:
df.InvoiceNo.value_counts()

573585    1114
581219     749
581492     731
580729     721
558475     705
          ... 
545341       1
545336       1
545331       1
573273       1
552135       1
Name: InvoiceNo, Length: 19960, dtype: int64

#### (五)、进行数据类型转换

In [30]:
df['CustomerID'] = df['CustomerID'].astype(str)
df['CustomerID']

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 524874, dtype: object

把`CustomerID`变量值结尾的`'.0'`给删除

In [31]:
df['CustomerID'] = df['CustomerID'].str.slice(0, 5)
df

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice CustomerID         Country  
0        12/1/2010 8:26       2.55      17850  United Kingdom  
1        12/1/2010 8:26       3.39      17850  United Kingdom  
2        12/1/2010 8:26       2.75      17850  United Kingdom  
3        12/1/2010 8:26       3.39      17850  United Kingdom  
4        12/1/2010 8:26       3.39      17850  United Kingdom  
...                 ...        ...        ...             ...  
541904  12/9/2011 12:50       0.85      12680          France  
541905  12/9/2011 12:50       2.10      12680          France  
541906  12/9/2011 12:50       4.15      12680          France  
541907  12/9/2011 12:50       4.15      12680          France  
541908  12/9/2011 12:50       4.95      12680          France  

[524874 rows x 8 columns]

In [32]:
df['Country'] = df['Country'].astype('category')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 524874 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype   
---  ------       --------------   -----   
 0   InvoiceNo    524874 non-null  object  
 1   StockCode    524874 non-null  object  
 2   Description  524874 non-null  object  
 3   Quantity     524874 non-null  int64   
 4   InvoiceDate  524874 non-null  object  
 5   UnitPrice    524874 non-null  float64 
 6   CustomerID   524874 non-null  object  
 7   Country      524874 non-null  category
dtypes: category(1), float64(1), int64(1), object(5)
memory usage: 32.5+ MB


把InvoiceDate变量的数据类型转换成日期时间类型

In [33]:
df['InvoiceDate'] = pd.to_datetime(df['InvoiceDate'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 524874 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    524874 non-null  object        
 1   StockCode    524874 non-null  object        
 2   Description  524874 non-null  object        
 3   Quantity     524874 non-null  int64         
 4   InvoiceDate  524874 non-null  datetime64[ns]
 5   UnitPrice    524874 non-null  float64       
 6   CustomerID   524874 non-null  object        
 7   Country      524874 non-null  category      
dtypes: category(1), datetime64[ns](1), float64(1), int64(1), object(4)
memory usage: 32.5+ MB


## 保存清理后的数据

In [34]:
df.to_csv('e_commerce_cleaned.csv', index=False)
pd.read_csv('e_commerce_cleaned.csv')

C:\Users\stube\AppData\Local\Temp\ipykernel_30172\4249193977.py:2: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv('e_commerce_cleaned.csv')


InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
524869    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
524870    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
524871    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
524872    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
524873    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

                InvoiceDate  UnitPrice  CustomerID         Country  
0       2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1       2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2       2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3       2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4       2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
...                     ...        ...         ...             ...  
524869  2011-12-09 12:50:00       0.85     12680.0          France  
524870  2011-12-09 12:50:00       2.10     12680.0          France  
524871  2011-12-09 12:50:00       4.15     12680.0          France  
524872  2011-12-09 12:50:00       4.15     12680.0          France  
524873  2011-12-09 12:50:00       4.95     12680.0          France  

[524874 rows x 8 columns]

输出df_cleaned时，有警告信息，同参考版一致